File: Spectrogeddon/Spectrogeddon/SettingsViewController.m
//
//  SettingsViewController.m
//  Spectrogeddon
//
//  Created by Tom York on 04/02/2015.
//  Copyright (c) 2015 Random. All rights reserved.
//

#import "SettingsViewController.h"
#import "SettingsWrapper.h"

@interface SettingsViewController ()
@end

@implementation SettingsViewController

@synthesize settingsModel = _settingsModel;

- (void)dealloc
{
    [NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(dismiss) object:nil];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

- (IBAction)changeColors:(id)sender
{
    [self.settingsModel nextColorMap];
    [self scheduleDismissal];
}

- (IBAction)toggleFrequencyStretch:(id)sender
{
    [self.settingsModel toggleFrequencyStretch];
    [self scheduleDismissal];
}

- (IBAction)changeScrollingSpeed:(id)sender
{
    [self.settingsModel nextScrollingSpeed];
    [self scheduleDismissal];
}

- (IBAction)changeSharpness:(id)sender
{
    [self.settingsModel nextSharpness];
    [self scheduleDismissal];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self scheduleDismissal];
}

- (void)scheduleDismissal
{
    [NSObject cancelPreviousPerformRequestsWithTarget:self selector:@selector(dismiss) object:nil];
    [self performSelector:@selector(dismiss) withObject:nil afterDelay:4.0];
}

- (void)dismiss
{
    [self performSegueWithIdentifier:@"dismissSettings" sender:self];
}

@end

File: Spectrogeddon/Spectrogeddon/SettingsViewController.swift
//
//  SettingsViewController.swift
//  Spectrogeddon
//
//  Created by Tom York on 12/05/2019.
//  Copyright © 2019 Random. All rights reserved.
//

import UIKit

class SettingsViewController : UIViewController {
    
    private var dismissalTimer: Timer?
    
    private var model: SettingsWrapper = SettingsWrapper()
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        dismissalTimer?.invalidate()
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        scheduleAutomaticDismissal()
    }

    @IBAction
    func changeColors() {
        settingsModel.nextColorMap()
        scheduleAutomaticDismissal()
    }

    @IBAction
    func toggleFrequencyStretch() {
        settingsModel.toggleFrequencyStretch()
        scheduleAutomaticDismissal()
    }

    @IBAction
    func changeScrollingSpeed() {
        settingsModel.nextScrollingSpeed()
        scheduleAutomaticDismissal()
    }
    
    @IBAction
    func changeSharpness() {
        settingsModel.nextSharpness()
        scheduleAutomaticDismissal()
    }
    
    func scheduleAutomaticDismissal() {
        dismissalTimer?.invalidate()
        dismissalTimer = Timer.scheduledTimer(withTimeInterval: 4, repeats: false) { [weak self] _ in
            

            Task { @MainActor in
                if let strongSelf = self {
                    strongSelf.performSegue(withIdentifier: "dismissSettings", sender: strongSelf)
                }
            }
        }
    }
}


extension SettingsViewController: SettingsModelClient {
    
    var settingsModel: SettingsWrapper {
        get {
            return model
        }
        set(settingsModel) {
            model = settingsModel
        }
    }
}
